data-prepper-plugins/failures-common/src/main/java/org/opensearch/dataprepper/plugins/dlq/s3/S3DlqWriter.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.dlq.s3;
+
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.Timer;
+import org.opensearch.dataprepper.plugins.dlq.DlqWriter;
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.configuration.DataPrepperVersion;
+import org.opensearch.dataprepper.model.failures.DlqObject;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ * S3 Dlq writer. Stores DLQ Objects in an S3 bucket.
+ *
+ * @since 2.2
+ */
+public class S3DlqWriter implements DlqWriter {
+
+    static final String S3_DLQ_RECORDS_SUCCESS = "dlqS3RecordsSuccess";
+    static final String S3_DLQ_RECORDS_FAILED = "dlqS3RecordsFailed";
+    static final String S3_DLQ_REQUEST_SUCCESS = "dlqS3RequestSuccess";
+    static final String S3_DLQ_REQUEST_FAILED = "dlqS3RequestFailed";
+    static final String S3_DLQ_REQUEST_LATENCY = "dlqS3RequestLatency";
+    static final String S3_DLQ_REQUEST_SIZE_BYTES = "dlqS3RequestSizeBytes";
+    private static final String KEY_NAME_FORMAT = "dlq-v%s-%s-%s-%s-%s";
+    private static final String FULL_KEY_FORMAT = "%s/%s";
```

Comment:
If they have a trailing path there will be an extra `/ `. The functionality still works it will just introduce an extra layer. I can add a validation to strip it but if someone already has a bucket with this prefix and wants to use it we would not be able to support it. I am inclined to leave it an provide documentation around it.